tests/utils/em.py

```diff
@@ -0,0 +1,313 @@
+    :type energy_model: EnergyModel
+
+    :params group_util: an utilization value for each group of cpus
+    :type group_util: [int]
+
+    :returns: a list for each group that contain a list of utilization per cpu.
+              For instance for a model that has four little cpus and two big
+              cpus it will return the following data:
+              [[group_util[0], 0, 0, 0], [group_util[1], 0]]
+    """
+    expected = []
+    for i, group in enumerate(energy_model.cpu_groups):
+        cpus_nb = len(group)
+        cpus_util = [0] * cpus_nb
+        cpus_util[0] = group_util[i]
+        expected.append(cpus_util)
+    return expected
+
+def ideal_placements(energy_model):
+    """
+    For each group, generates a workload that should run only on it and
+    controls that obtained placement is equal to the expected placement
+    :params energy_model: The energy model for which these workloads are
+                          generated
+    :type energy_model: EnergyModel
+
+    :returns: True if the expected placement is equal to the obtained one,
+              False otherwise
+    """
+    msg = ''
+    avg_cap = get_avg_cap(energy_model)
+    for i, group in enumerate(energy_model.cpu_groups):
+        one_task = {'t0': avg_cap[i]}
+        placement = energy_model.get_optimal_placements(one_task)
+        exp_cap = [0] * len(energy_model.cpu_groups)
+        exp_cap[i] = avg_cap[i]
+        expected = _get_expected_placement(energy_model, exp_cap)
+        if not _are_placements_equal(energy_model, expected, placement):
+            msg = ('The expected placement for the task {} is {} but the '
+                    'obtained placement \n\t\t'
+                    'was {}'.format(one_task, expected, placement))
+            return (False, msg)
+    return (True, msg)
+
+def _get_efficiency(util, cpu):
+    """
+    For a given utilization, computes the energy efficiency (capacity / power)
+    to run at this utilization. If it does not correspond to an opp for the
+    cpu, the efficiency is computed by selecting the opp above the
+    utilization and the power is computed proportionally to it.
+    :params util: targeted utilization
+    :type util: int
+
+    :params cpu: The cpu for which the energy efficiency is computed.
+    :type cpu: EnergyModelNode
+
+    :returns: The ratio between the capacity and the power for the targeted opp
+    """
+
+    # Get the list of capacities and powers for the cpu
+    capacities = _get_capacities_list(cpu)
+    powers = _get_powers_list(cpu)
+
+    # Compute the efficiency for the first capacity larger than the
+    # requiered opp
+    for cap in capacities:
+        if cap >= util:
+            power = powers[capacities.index(cap)] * (float(util) / cap)
+            return float(cap) / power
+
+    raise ValueError('The utilization {} is larger than the possible '
+                     'opp for {}'.format(util, cpu.name))
+
+def check_overutilized_area(energy_model):
```

Comment:
All of the other tests look fine to me, but this one got me slightly confused.
IIUC, in short you want to assert here that the biggest OPPs on the LITTLEs are still more energy efficient than the smallest OPPs on the bigs, and that we don't skip those because of the 80% margin.

Maybe that should actually be split into two tests, i.e:
1) Is there any OPP on the bigs that is more power-efficient than on the LITTLEs ?
2) Is there any OPP on the LITTLEs that is more power-efficient that on the bigs but we'll skip it because of that 80% margin ?

What do you think ?

Reply:
Thanks for your comment
Yes it seems better that way :) I will work on it